Keep the cursor position in the Classic Block

azaozz opened the pull request:
Fixes #10509.

- On blur store the cursor position and the "raw" content of the Classic Block.
- Use the "raw" content when refreshing the editor content on `componentDidUpdate()`. It contains the cursor position "bookmark".
- Restore the cursor position on focus and before inserting content.

azaozz committed:
Keep the cursor position on blur

- On blur store the cursor position and the "raw" content of the Classic Block.
- Use the "raw"  content when refreshing the editor content on `componentDidUpdate()`.
- Restore the cursor position on focus and before inserting content.

youknowriad commented on packages/block-library/src/classic/edit.js:
```diff
@@ -50,12 +50,20 @@ export default class ClassicEdit extends Component {
-
+		const { clientId, attributes } = this.props;
 		const editor = window.tinymce.get( `editor-${ clientId }` );
+		let { contentRaw } = attributes;
```
contentRaw is not persisted, we should probably use "local state" for it.

azaozz replied:
Ah, sure, that'd work even better :)

youknowriad reviewed (commented):
We could use some e2e test for the classic block :)

azaozz committed:
Store the initial "raw" content on editor init

Needed when the first thing the user does after adding a Classic Block is to open a modal that inserts something, like the media modal.

iseulde commented on packages/block-library/src/classic/edit.js:
```diff
@@ -84,12 +92,29 @@ export default class ClassicEdit extends Component {
+
+			// Keep the "raw" content as it contains the caret position bookmark.
 			setAttributes( {
+				contentRaw: editor.getContent( { format: 'raw' } ),
```
Storing the raw content without any processing seems like a bad idea? Why should this be different from the classic editor?

iseulde commented on packages/block-library/src/classic/edit.js:
```diff
@@ -84,12 +92,29 @@ export default class ClassicEdit extends Component {
+		editor.on( 'init', () => {
+			// Store the initial "raw" content.
+			setAttributes( {
+				contentRaw: editor.getContent( { format: 'raw' } ),
```
I wouldn't expect merely initialising the block to change the content. The user didn't do  anything.

iseulde commented on packages/block-library/src/classic/edit.js:
```diff
@@ -50,12 +50,20 @@ export default class ClassicEdit extends Component {
 		const editor = window.tinymce.get( `editor-${ clientId }` );
+		let { contentRaw } = attributes;
+
+		// Strip white space before comparing to avoid resetting the editor content when not necessary;
```
What kind of whitespace is different?

azaozz replied:
The only differences I'm seeing are in some `<p>\u00a0</p>` and some `;` missing/being added to ends of inline style attributes.

azaozz commented:
Looking more at this, still can't figure out what the refresh code in `componentDidUpdate()` is supposed to do:
```
const { clientId, attributes: { content } } = this.props;
....
if ( prevProps.attributes.content !== content ) {
	editor.setContent( content || '' );
}
```

Seems at some point it was possible that `attributes.contend` was changed/updated outside the editor, but that doesn't seem to be the case now. Perhaps this is a "left over" from how undo/redo was working previously?

Even if this is still needed, don't think that looking at `prevProps.attributes.content` is a good way to determine what's in the editor at that moment. It needs to do `editor.getContent()`.

@iseulde @aduth any chance you could remember what was the purpose of the above code? I know, it's been there for over a year...

iseulde commented:
To me, this is an interesting thing that happens in master:

1. Load a classic block with some text. Place the caret somewhere other than the start.
2. Click the media button in the toolbar.
3. Select an image.
4. Close the media modal. Do not insert it!
5. Place the caret back where it was (somewhere other than the very start).
6. Click the media button in the toolbar.
7. The image you selected last time will already be selected. Do not click it again!
8. Insert the already selected image.
9. The image is correctly inserted.

This means that clicking the image specifically is causing the focus loss?

iseulde commented:
> Seems at some point it was possible that attributes.contend was changed/updated outside the editor, but that doesn't seem to be the case now.

I think it is still needed for undo/redo. It needs to be possible to set the content from outside the classic block.

> Even if this is still needed, don't think that looking at prevProps.attributes.content is a good way to determine what's in the editor at that moment. It needs to do editor.getContent().

Sounds good to me.

azaozz commented:
> This means that clicking the image specifically is causing the focus loss?

Yes, clicking anywhere outside the editor triggers the `blur()` and looses the caret position in the editor body <div>.

iseulde commented:
Another interesting thing: even if I remove `componentDidUpdate` entirely, the problem still exists. Again only happens if you select the image in the media library. Otherwise note.

azaozz commented:
> even if I remove componentDidUpdate entirely, the problem still exists

Right. It still needs `getBookmark()`, `moveToBookmark()` on blur/focus.

iseulde commented:
> Right. It still needs getBookmark(), moveToBookmark() on blur/focus.

Why does this only happen for selecting an image though? What's special about this image?

azaozz commented:
Here it happens for any click that focuses something outside the editor body. For example opening the menu, etc.

azaozz commented:
Found it :) 

The check in componentDidUpdate() is needed only when the global undo/redo buttons are clicked. Not needed/not running for Ctrl/Cmd + z though...

azaozz commented:
Closing this in favor of #12415.
